agent/consul/state/intention.go

```diff
@@ -373,6 +353,63 @@ func (s *Store) IntentionMatch(ws memdb.WatchSet, args *structs.IntentionQueryMa
 	return idx, results, nil
 }
 
+// IntentionMatchOne returns the list of intentions that match the namespace and
+// name for a single source or destination. This applies the resolution rules
+// so wildcards will match any value.
+//
+// The returned intentions are sorted based on the intention precedence rules.
+// i.e. result[0] is the highest precedent rule to match
+func (s *Store) IntentionMatchOne(ws memdb.WatchSet,
+	entry structs.IntentionMatchEntry, matchType structs.IntentionMatchType) (uint64, structs.Intentions, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	// Get the table index.
+	idx := maxIndexTxn(tx, intentionsTableName)
+	if idx < 1 {
+		idx = 1
+	}
+
+	results, err := s.intentionMatchOneTxn(tx, ws, entry, matchType)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	sort.Sort(structs.IntentionPrecedenceSorter(results))
+
+	return idx, results, nil
+}
+
+func (s *Store) intentionMatchOneTxn(tx *txn, ws memdb.WatchSet,
```

Comment:
We've been changing over all these `Store` methods that don't use their method receiver to be functions, so that they don't require a `Store` to be used. In this case I guess `intentionMatchGetParams` also does not need the `Store` method receiver. I think if you make that one a function we could remove the `Store` method receiver here as well.

We're also moving to use `ReadTxn` in place of `*txn` so that we can reduce some of the extra wrapping we have around `memdb.MemDB`, and also so that these function signatures are a little more explicit about if they perform reads or writes.

Reply:
I like that, just updated it!